README.md
# hiera_redis

#### Table of Contents

1. [Description](#description)
2. [Setup - The basics of getting started with hiera_redis](#setup)
    * [Setup requirements](#setup-requirements)
    * [Beginning with hiera_redis](#beginning-with-hiera_redis)
3. [Usage - Configuration options and additional functionality](#usage)
4. [Limitations - OS compatibility, etc.](#limitations)
5. [Development - Guide for contributing to the module](#development)
6. [Credits - credits to developers](#credits)

## Description

This module provides a Hiera 5 backend for Redis.
With support recieve from redis hash instead String only. JSON requered for this module.
Also support interpolate results from redis - patterns in result like `%{hiera('somekey')}` or  `%{lookup('somekey')}`  will be replaced to `somekey` value from Hiera.

## Setup

### Setup Requirements

The backend requires the [redis](https://github.com/redis/redis-rb) gem installed in the Puppet Server JRuby.
It can be installed with:

    /opt/puppetlabs/bin/puppetserver gem install redis

It is also recommended to install the gem into the agent's Ruby:

    /opt/puppetlabs/puppet/bin/gem install redis

This allows commands such as `puppet apply` or `puppet lookup` to use the backend.

Also need to install `json` gem

### Beginning with hiera_redis

If Redis is running on the Puppet master with the default settings, specifying the `lookup_key` as 'redis_lookup_key' is sufficient, for example:

```yaml
---
version: 5
hierarchy:
  - name: hiera_redis
    lookup_key: redis_lookup_key
```

## Usage

By default, the backend will query Redis with the key provided.
It is also possible to query multiple scopes such as with the YAML backend, where the expected key in Redis is composed of the scope and the key separated by a character (default is `:`). For example, the following can be used:

```yaml
---
version: 5
hierarchy:
  - name: hiera_redis
    lookup_key: redis_lookup_key
    options:
      confine_to_keys:
        - '^redis_.*'
        - '^myapp_.*'
        - '^ssh_group$'
      scopes:
        - "osfamily/%{facts.os.family}"
        - common
```

The backend then expects keys of a format such as `common:foo::bar` for a lookup of 'foo::bar'.

The other options available include:

* `host`: The host that Redis is located on. Defaults to 'localhost'.
* `port`: The port that Redis is running on. Defaults to 6379.
* `socket`: Optional Unix socket path
* `password`: Optional Redis password
* `db`: The database number to query on the Redis instance. Defaults to 0.
* `scope`: The scope to use when querying the database.
* `scopes`: An array of scopes to query. Cannot be used in conjunction with the `scope` option.
* `separator`: The character separator between the scope and key being queried. Defaults to ':'.
* `confine_to_keys`: Only use this backend if the key matches one of the regexes in the array.
* `connect_timeout`: connect timeout to redis, seconds. Default 0.5
* `read_timeout`: read timeout for redis, seconds. Default 0.5
* `write_timeout`: write timeout for redis, seconds. Default 0.5

### confine_to_keys config example:

```yaml
  confine_to_keys:
    - "application.*"
    - "apache::.*"
```

## Limitations

Tested partly for puppet5, use for you own risk.

## Credits

This repository was initially a fork from [maxadamo hiera_redis](https://github.com/maxadamo/hiera_redis) repo, forked for add some fix.

The code to related to add supporting of getting hashes also from hiera redis storage. For now if in recieved result from redis we have "{" and "}" symvols, we will change type of result as HASH and puppet will be use result from redis as hash.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## Release 0.3.0 

** Bugfixes **

* Added `redis.close()` action for closing redis connection, without this, we probably will have issues with free TCP sockets on puppetserver

** Features **

* Added timeouts to redis connections, also added options to configure it, see more in README. by default all timeouts 0.5 seconds(read,write,connect).


## Release 0.2.0


** Bugfixes**

* If we will find `{` and `}` in result from redis, we will need to make it hash - this part always recieve `{}` e.g. empty hash. - For now we use json for make hash from string, if it is posable
* Interpolation not working - for now if result from redis has some what puppet can interpolate(get value from hiera), e.g. some like `%{hiera('somekey')}`, it is will be replaced by `somekey` value




## Release 0.1.0

**Features**

* added condition for detect hash in data into redis key and change result type to HASH type for puppet.

**Bugfixes**

**Known Issues**


lib/puppet/functions/redis_lookup_key.rb
Puppet::Functions.create_function(:redis_lookup_key) do
  begin
    require 'redis'
    require 'json'

  rescue LoadError
    raise Puppet::DataBinding::LookupError, 'The redis and json  gem must be installed to use redis_lookup_key'
  end

  dispatch :redis_lookup_key do
    param 'Variant[String, Numeric]', :key
    param 'Hash', :options
    param 'Puppet::LookupContext', :context
  end

  def redis_lookup_key(key, options, context)
    return context.cached_value(key) if context.cache_has_key(key)

    if (confine_keys = options['confine_to_keys'])
      raise ArgumentError, '[hiera-redis] confine_to_keys must be an array' unless confine_keys.is_a?(Array)

      begin
        confine_keys = confine_keys.map { |r| Regexp.new(r) }
      rescue StandardError => e
        raise Puppet::DataBinding::LookupError, "[hiera-redis] creating regexp failed with: #{e}"
      end

      regex_key_match = Regexp.union(confine_keys)

      unless key[regex_key_match] == key
        context.explain { "[hiera-redis] Skipping hiera_redis backend because key '#{key}' does not match confine_to_keys" }
        context.not_found
      end
    end

    host      = options['host']      || 'localhost'
    port      = options['port']      || 6379
    socket    = options['socket']    || nil
    password  = options['password']  || nil
    db        = options['db']        || 0
    scopes    = options['scopes']    || [options['scope']]
    separator = options['separator'] || ':'
# timeout options, by default 0.5 seconds
    connect_timeout = options['connect_timeout'] || 0.5
    read_timeout = options['read_timeout'] || 0.5
    write_timeout = options['write_timeout'] || 0.5
# added timeouts for redis connections
# without it, we will get a lot of not closed TCP connections
     redis = if !socket.nil? && !password.nil?
              Redis.new(path: socket, password: password, db: db, connect_timeout: connect_timeout, read_timeout: read_timeout, write_timeout: write_timeout)
            elsif !socket.nil? && password.nil?
              Redis.new(path: socket, db: db, connect_timeout: connect_timeout, read_timeout: read_timeout, write_timeout: write_timeout)
            elsif socket.nil? && !password.nil?
              Redis.new(password: password, host: host, port: port, db: db, connect_timeout: connect_timeout, read_timeout: read_timeout, write_timeout: write_timeout)
            else
              Redis.new(host: host, port: port, db: db, connect_timeout: connect_timeout, read_timeout: read_timeout, write_timeout: write_timeout)
            end
    result = nil

    scopes.each do |scope|
      redis_key = scope.nil? ? key : [scope, key].join(separator)
      result = redis_get(redis, redis_key)

      break unless result.nil?
    end
# close redis connection, for fix issue with TCP connects
    redis.close()

    context.not_found if result.nil?
    # if result contains some hiera or lookup pattern for interpolate it, we need try to make it
    #to interpolate some subincluded lookups neet interpolate result in first.
    if (result.include? "%{hiera") || (result.include? "%{lookup")
      result =  context.interpolate(result)
    end
    # if we can validate json in result, we need to make it and change type of result for context.
    if valid_json(result)
       context.cache(key, Hash(JSON.parse(result)))
    else
       context.cache(key, result)
    end
  end
# we just trt validate it, if we can without error - return true, if not - false.
  def valid_json(json)
      JSON.parse(json)
      return true
  rescue TypeError, JSON::ParserError => e
     return false
  end

  def redis_get(redis, key)
    case redis.type(key)
    when 'string'
      redis.get(key)
    when 'list'
      redis.lrange(key, 0, -1)
    when 'set'
      redis.smembers(key)
    when 'zset'
      redis.zrange(key, 0, -1)
    when 'hash'
      redis.hgetall(key)
    end
  end
end

